### PR TITLE
CLOUDP-175080: Deletion protection for project resources (final)

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

**Review by igor-karpukhin — APPROVED** [review]
Massive work here! Good job! 👍

File: pkg/controller/atlasproject/auditing.go [hunk 1]
```diff
@@ -59,12 +81,24 @@ func auditingInSync(atlas *mongodbatlas.Auditing, spec *v1.Auditing) bool {
...
+	if !isAuditingEmpty(spec) {
+		specAsAtlas = spec.ToAtlas()
+	}
+
+	if isAuditingEmpty(atlas) {
+		atlas = &mongodbatlas.Auditing{
+			AuditAuthorizationSuccess: toptr.MakePtr(false),
+			Enabled:                   toptr.MakePtr(false),
+		}
 	}
```
Comment (josvazg) [inline pkg/controller/atlasproject/auditing.go:98]: I do not understand this change, can you elaborate?

Comment (helderjs) [reply pkg/controller/atlasproject/auditing.go:98]: While unlikely to happen, this is a protection once the response from the client is a pointer. Given that, if for any reason it returns no error and nil as a response, we assume default values.

File: pkg/controller/atlasproject/maintenancewindow.go [hunk 2]
```diff
@@ -169,3 +193,65 @@ func toggleAutoDeferInAtlas(client mongodbatlas.Client, projectID string) workfl
...
+
+	if akoMWindow.StartASAP != *atlasMWindow.StartASAP {
+		return false
+	}
+
+	if akoMWindow.AutoDefer != *atlasMWindow.AutoDeferOnceEnabled {
+		return false
+	}
+
+	return true
```
Comment (josvazg) [inline pkg/controller/atlasproject/maintenancewindow.go:252]: Why not this instead?

```suggestion
	return akoMWindow.DayOfWeek == atlasMWindow.DayOfWeek &&
		akoMWindow.HourOfDay == *atlasMWindow.HourOfDay &&
		akoMWindow.StartASAP != *atlasMWindow.StartASAP &&
		akoMWindow.AutoDefer != *atlasMWindow.AutoDeferOnceEnabled
```

For me it is easier to read as it separates the pointer value fixes above from the actual equals check.

File: pkg/controller/atlasproject/maintenancewindow_test.go [hunk 3]
```diff
@@ -1,13 +1,50 @@
...
+type maintenanceWindowClient struct {
+	GetFunc func(projectID string) (*mongodbatlas.MaintenanceWindow, *mongodbatlas.Response, error)
+}
+
+func (c *maintenanceWindowClient) Get(_ context.Context, projectID string) (*mongodbatlas.MaintenanceWindow, *mongodbatlas.Response, error) {
+	return c.GetFunc(projectID)
+}
+
+func (c *maintenanceWindowClient) Update(_ context.Context, _ string, _ *mongodbatlas.MaintenanceWindow) (*mongodbatlas.Response, error) {
+	return nil, nil
```
Comment (josvazg) [inline pkg/controller/atlasproject/maintenancewindow_test.go:33]: You can use [impl](https://github.com/josharian/impl) which will by default panic unimplemented here. This is what we recommend in (docs/testing.go)[https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/docs/testing.md]

Not a big deal, but the more similar these mocks are, the less refactor they will need when we move them to their own package.

Comment (helderjs) [reply pkg/controller/atlasproject/maintenancewindow_test.go:33]: Here I think it's not necessary to apply this once there's no one else consuming these mocks. Next week when I'm back I'll be addressing the ticket to move all the mocks to its own pkg. Also, I will provide "full-featured" mocks together.

File: pkg/controller/atlasproject/teams.go [hunk 4]
```diff
@@ -243,3 +267,85 @@ func hasTeamRolesChanged(current []string, desired []v1.TeamRole) bool {
...
+		if atlasAssignedTeam != nil {
+			atlasAssignedTeamsInfo = append(
+				atlasAssignedTeamsInfo,
+				assignedTeamInfo{
+					ID:    atlasAssignedTeam.TeamID,
+					Roles: atlasAssignedTeam.RoleNames,
+				},
+			)
+		}
+	}
```
Comment (josvazg) [inline pkg/controller/atlasproject/teams.go:309]: nitpick
This seems to be a format conversion. Moving to its own function could made this more readable.

Comment (josvazg) [inline pkg/controller/atlasproject/teams.go:309]: nitpick
This seems to be a format conversion. Moving to its own function could made this more readable.

```suggestion
	atlasAssignedTeamsInfo := atlasAssignedTeamsToK8s(atlasAssignedTeams.Results)
```

**Review by josvazg — APPROVED** [review]
LGTM

All my comments are non blocking nitpicks

**Pull request merged**